### PR TITLE
Rz/pnl calcs

### PR DESCRIPTION
**Pull request by pacxiu**
# [P&L calculations](https://app.shortcut.com/oazo-apps/story/2565/creation-of-function-calculating-p-l-as-a-single-number-using-extended-vault-history)

  
## Changes 👷‍♀️
- Add function to calculate P&L based on events 

  
## How to test 🧪
- Compare test result with calculation sheet
 https://docs.google.com/spreadsheets/d/1VnPUV0x0ujWdIc-1OvqrCX--KwSBgHSb6ata8wbr69g/edit#gid=0
    
## Definition of done ✔️

- [ ] Acceptance criteria for each issue met
- [ ] Unit tests written where needed and passing
- [ ] End-to-end tests for happy path
- [ ] Documentation updated <When applicable>
- [ ] Project builds without errors
- [ ] Non-functional requirements met
- [ ] Code reviewed and functionality tested by the reviewer (locally, Heroku, etc.)
- [ ] Feature verified and accepted by Product Owner
- [ ] Project deployed to production environment

### CODE REVIEW

File: helpers/multiply/calculations.ts [hunk 1]
```diff
@@ -234,3 +236,53 @@ export function calculateCloseToCollateralParams(
...
+      const { generated } = event
+
+      cummulativeWithdrawnUSD = cummulativeWithdrawnUSD.plus(generated)
+    }
+
+    cummulativeFeesUSD = cummulativeFeesUSD.plus(event.gasFee.times(event.marketPrice))
+  })
+
+  // console.log(`
+  //   starting ${cummulativeDepositUSD.minus(cummulativeWithdrawnUSD).toFixed()}
```
Comment (peculiarity) [inline helpers/multiply/calculations.ts:279]: Remove this please

File: helpers/multiply/calculations.ts [hunk 2]
```diff
@@ -234,3 +235,74 @@ export function calculateCloseToCollateralParams(
...
+}
+
+function getCumulativeWithdrawnUSD(total: BigNumber, event: VaultEvent) {
+  switch (event.kind) {
+    case 'WITHDRAW':
+    case 'WITHDRAW-PAYBACK':
+      return total.plus(event.collateralAmount.abs().times(event.oraclePrice))
+    case 'GENERATE':
+    case 'DEPOSIT-GENERATE':
+      return total.plus(event.daiAmount)
```
Comment (adamskrodzki) [inline helpers/multiply/calculations.ts:267]: is sign here correct ? no abs() and - required ?

File: helpers/multiply/calculations.ts [hunk 3]
```diff
@@ -234,3 +235,74 @@ export function calculateCloseToCollateralParams(
     loanFee: currentDebt.times(FF),
   }
 }
+
+function getCumulativeDepositUSD(total: BigNumber, event: VaultEvent) {
```
Comment (adamskrodzki) [inline helpers/multiply/calculations.ts:239]: where this methods are in use ?

**Review by adamskrodzki — CHANGES_REQUESTED** [review]

File: helpers/multiply/calculations.ts (hunk 2 above)
Comment (paszkowskiDamian) [reply helpers/multiply/calculations.ts:267]: Here `daiAmount` will always be positive as we increase the debt.

File: helpers/multiply/calculations.ts (hunk 3 above)
Comment (paszkowskiDamian) [reply helpers/multiply/calculations.ts:239]:
```
export function calculatePNL(events: VaultEvent[], currentNetValueUSD: BigNumber) {
  const cumulativeDepositUSD = events.reduce(getCumulativeDepositUSD, zero) // here line 299
  const cumulativeWithdrawnUSD = events.reduce(getCumulativeWithdrawnUSD, zero)
  const cumulativeFeesUSD = events.reduce(getCumulativeFeesUSD, zero)

  return cumulativeWithdrawnUSD
    .plus(currentNetValueUSD)
    .minus(cumulativeFeesUSD)
    .minus(cumulativeDepositUSD)
    .div(cumulativeDepositUSD)
}
```

File: helpers/multiply/calculations.ts [hunk 4]
```diff
@@ -234,3 +235,74 @@ export function calculateCloseToCollateralParams(
...
+    case 'CLOSE_VAULT_TO_COLLATERAL':
+      return total.plus(event.exitCollateral.times(event.marketPrice))
+    case 'CLOSE_VAULT_TO_DAI':
+      return total.plus(event.exitDai)
+    case 'AUCTION_STARTED':
+    case 'AUCTION_STARTED_V2':
+      return zero
+    default:
+      return total
+    // TODO: move_src
```
Comment (paszkowskiDamian) [inline helpers/multiply/calculations.ts:281]: this requires changes in cache

File: helpers/multiply/calculations.ts [hunk 5]
```diff
@@ -234,3 +235,74 @@ export function calculateCloseToCollateralParams(
...
+}
+
+function getCumulativeFeesUSD(total: BigNumber, event: VaultEvent) {
+  switch (event.kind) {
+    case 'OPEN_MULTIPLY_VAULT':
+    case 'DECREASE_MULTIPLE':
+    case 'INCREASE_MULTIPLE':
+    case 'CLOSE_VAULT_TO_COLLATERAL':
+    case 'CLOSE_VAULT_TO_DAI':
+      return total.plus(event.gasFee.times(event.marketPrice)) // TODO add ethPrice to all events
```
Comment (paszkowskiDamian) [inline helpers/multiply/calculations.ts:292]: this requires change in cache

File: helpers/multiply/calculations.ts [hunk 6]
```diff
@@ -234,3 +235,74 @@ export function calculateCloseToCollateralParams(
...
+      return zero
+    default:
+      return total
+    // TODO: move_src
+  }
+}
+
+function getCumulativeFeesUSD(total: BigNumber, event: VaultEvent) {
+  switch (event.kind) {
+    case 'OPEN_MULTIPLY_VAULT':
```
Comment (adamskrodzki) [inline helpers/multiply/calculations.ts:287]: only Gas Fees for multiply ? we do not have one for not multiply events ?

**Review by adamskrodzki — APPROVED** [review]

**Pull request merged**